server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -760,7 +772,10 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
                     maybeUpdatedState = ClusterState.builder(maybeUpdatedState).metadata(metadataBuilder).build();
                 }
 
+                maybeUpdatedState = maybeRemoveIndexRefreshBlocks(maybeUpdatedState, indicesWithUnpromotableShardsStarted);
```

Comment:
I am fine with this going in for now, it is a good improvement as is.

I wonder though if we should follow-up with a change to remove the refresh block in a subsequent cluster state update. That would avoid problems with a coordinator not seeing the shard started message but an indexing node seeing the refresh block removed, which I *think* could cause a refresh to be ack'ed and a subsequent search not seeing the refreshed data?

Reply:
> That would avoid problems with a coordinator not seeing the shard started message but an indexing node seeing the refresh block removed, which I think could cause a refresh to be ack'ed and a subsequent search not seeing the refreshed data?

I believe that there's the same risk with the deferred removal, right? The cluster state update with the unblocked index might be applied a bit later on the search node anyway, unless you're proposing to react to the shard started events differently on the search nodes?

Reply:
Also, the shard is marked as started as soon as it is ready to serve searches, so that should be fine too?

Reply:
On the first part, I would rely on cluster state updates trying hard to be applied everywhere before proceeding with the next update. This means we would be sure to see the shard started everywhere before the refresh block is removed.

Implicit in that is that I thought we would then forward search requests always when there is a started shard.

Reply:
> we could send the shard search request to the initializing shard, which could then make the determination. I think that would be safe too.

This would come with the downside of risking sending search requests to search nodes where the shard is still unassigned. I think that is possible to handle.

But it seems there are two options and the trade-off is not entirely clear.

(but as mentioned I am ok with doing what is in this PR for now and then let us tackle that other piece).

Reply:
> (but as mentioned I am ok with doing what is in this PR for now and then let us tackle that other piece).

Sounds good, I'll open a follow up ticket for this.